pywren/wait.py

```diff
@@ -81,10 +82,47 @@ def _wait(fs, THREADPOOL_SIZE):
 
     storage_config = wrenconfig.extract_storage_config(wrenconfig.default())
     storage_handler = storage.Storage(storage_config)
-    callids_done = storage_handler.get_callset_status(callset_id)
 
+    # Signal the completion of tasks by probing status files.
+    # Because S3 does not provide list-after-write consistency (which might also apply
+    # to other storage backend). List() can be only used as an optimization
+    # but not a timely way to signal completion. Thus, our strategy is to:
+    # 1) do list()
+    # 2) use get() to signal N tasks that do not show up in 1)
+    # 3) repeat 2) if all N tasks completed, otherwise stop
+    # Note: a small N is probably preferred here.
+
+
+    callids_done = storage_handler.get_callset_status(callset_id)
     callids_done = set(callids_done)
 
+    num_samples = 4
+    still_not_done_futures = [f for f in not_done_futures if (f.call_id not in callids_done)]
+    def fetch_status(f):
+        return storage_handler.get_call_status(f.callset_id, f.call_id)
+
+    pool = ThreadPool(num_samples)
+    # repeat util all futures are done
+    while still_not_done_futures:
```

Comment:
Is this actually the design contract of `_wait`? I thought it would perform _one pass_ of checking if things are done and downloading those; it would not block waiting for all to finish (note the original `_wait` has no looping, that's handeled in `wait()`)